torchchat/model.py

```diff
@@ -343,13 +413,26 @@ def _unfuse_wqkv_state_dict(
 
         _unfuse_wqkv_state_dict(state_dict, self.dim)
 
+    def distribute(self, device_mesh: DeviceMesh):
+        self.device_mesh = device_mesh
+        parallelize_module(self.wq, device_mesh, ColwiseParallel())
+        parallelize_module(self.wk, device_mesh, ColwiseParallel())
+        parallelize_module(self.wv, device_mesh, ColwiseParallel())
+        parallelize_module(self.wo, device_mesh, RowwiseParallel(output_layouts=Shard(1)))
+        # TODO: enable kv cache in distributed case
+        self.kv_cache = None
+
     def forward(
         self,
         x: Tensor,
         freqs_cis: Tensor,
         mask: Tensor,
         input_pos: Optional[Tensor] = None,
     ) -> Tensor:
+        # Gather sequence back in case of sequence parallelism before attention
+        if isinstance(x, DTensor):
```

Comment:
minor perf q:
could Transformer provide:
self.seq_parallel_degree
to TransformerBlock.Attention during init 
so that this isinstance(x) check doesn't have to be done on every forward?
Instead it would just be:
if self.seq_parallel > 1:

Reply:
Good question on perf. 
On the other hand, seq_parallel > 1 does not necessarily mean x is a `DTensor` (which depends on whether `to_local` has been called), so calling `.redistributed` on it is an uncertain behavior.

Reply:
good point that x may not always be dtensor. 
Just for kicks, made a quick script to compare perf - it's not that big (~5%):
~~~
Rank 0:
Time taken for isinstance check: 0.004911 seconds
Time taken for int setting check: 0.004534 seconds
isinstance is 1.083x slower than int setting check
Rank 1:
Time taken for isinstance check: 0.004674 seconds
Time taken for int setting check: 0.004502 seconds
isinstance is 1.038x slower than int setting check
~~~